### PR TITLE
Lighten and Optimize Build CI

### PR DESCRIPTION
**Pull request by rashidnhm**
**Context:**
Update the CI to run a "lightened" version of tests by default on pull requests, and increase the number of groups for Jax and TF tests.

**Description of the Change:**
This PR introduces the following changes:

- A 'ligtened' test suite
    - Instead of running the test suite across multiple Python versions (3.9, 3.10, 3.11), only run the test suite against one version of Python (3.9).

- Reduce max parallelisation for small test jobs, and assign the extra compute room to Jax and TF tests.

- Introduction of "test durations" files to be used by pytest-split in order to better distribute jobs across groups.
    - These two files add an additional 3.1MB. At the time of writing these two files are included as part of this PR, but it can be external and sourced as part of CI.
    - These files DO NOT need to be regularly updated, even if new JAX and TF tests are added, as Pytest takes an avg time for tests that are not in the durations file. If we could update them once in a while (maybe once a year or two), that should suffice.

**Benefits:**
The changes introduced as part of this PR reduces the runtime of the build CI. As well as consume less CI workers for PRs (by default).

Full Test Suite (Multiple Versions of Python):
Runtime:
Before: ~45min
Now:    ~25min

Workers:
Before: 30
Now:    29

Lightened Test Suite (One version of Python):
Runtime:
Before: N/A
Now:    ~15min

Workers:
Before: N/A
Now:    17

Overall, the CI on a PR that is doing a lightened test run will now consume 43% less workers, and run 67% quicker.

**Possible Drawbacks:**
Not running multiple versions of Python on PRs does open the door for regressions. However, if the label `ci:run-full-test-suite` is attached to any PR, then the entire test suite runs on it. All merges to master also run the full test suite.

**Related GitHub Issues:**
None.

### CODE REVIEW

File: tests/pulse/test_convenience_functions.py
Comment (rashidnhm) [reply tests/pulse/test_convenience_functions.py:193]: These changes I pulled in from #4643 which sped up certain tests.

[See Thread Here](https://xanaduhq.slack.com/archives/C03BY2J49HT/p1696410763746179?thread_ts=1696367684.264179&cid=C03BY2J49HT)

Comment (albi3ro) [reply tests/pulse/test_convenience_functions.py:193]: thanks 👍

**Review by albi3ro — APPROVED** [review]
We will need to see if the decreased code coverage due to not running data and qchem tests cause problems, but we can adjust as needs be.

I will freely admit I don't understand how some of the CI specification stuff works, but it looks substantially faster. If I need to figure out how to solve a particular problem later on, I will figure out how to solve the relevant problem then.

Are you able to change the required CI checks for the repo? We should probably do so right before this gets merged in.

**Comment by trbromley** [comment]
> due to not running data and qchem tests

Are we doing that in this PR? i.e., not running them on PRs?

**Comment by rashidnhm** [comment]
> > due to not running data and qchem tests
> 
> Are we doing that in this PR? i.e., not running them on PRs?

Yes! You can go to the 'Checks' tab to see skipped tests
image


We can specify which jobs to skip in PRs as part of [this variable](https://github.com/PennyLaneAI/pennylane/blob/0a671250b1a7698a4c80f8b4eed48a4171f4a04d/.github/workflows/tests.yml#L30). It is configurable to skip whichever jobs we'd want.

**Comment by rashidnhm** [comment]
> We will need to see if the decreased code coverage due to not running data and qchem tests cause problems, but we can adjust as needs be.
> 
> I will freely admit I don't understand how some of the CI specification stuff works, but it looks substantially faster. If I need to figure out how to solve a particular problem later on, I will figure out how to solve the relevant problem then.
> 
> Are you able to change the required CI checks for the repo? We should probably do so right before this gets merged in.

Are there particular things you would like documented? I can add a README to .github/workflows that can outline high or mid level how the CI workflows go.

**Comment by trbromley** [comment]
> > > due to not running data and qchem tests
> > 
> > 
> > Are we doing that in this PR? i.e., not running them on PRs?
> 
> Yes! You can go to the 'Checks' tab to see skipped tests image
> 
> We can specify which jobs to skip in PRs as part of [this variable](https://github.com/PennyLaneAI/pennylane/blob/0a671250b1a7698a4c80f8b4eed48a4171f4a04d/.github/workflows/tests.yml#L30). It is configurable to skip whichever jobs we'd want.

Thanks @rashidnhm! How much do we save by skipping? I'm wondering whether it is worth keeping these tests.

**Comment by rashidnhm** [comment]
> Thanks @rashidnhm! How much do we save by skipping? I'm wondering whether it is worth keeping these tests.

We skip these and assign the workers to Jax/TF tests, which is where some of the time savings happen. But if we are okay with using more workers for lightened test runs, we can easily re-enable them. Let me know.

File: .github/workflows/interface-unit-tests.yml
Comment (timmysilv) [inline .github/workflows/interface-unit-tests.yml:84]: is this the best way to save dicts between job stages in github CI? I mean.. it gets the job (sorry) done lol

**Review by timmysilv — APPROVED** [review]
yaml and github actions are always a little strange so I am double-grateful that you did this. all looks good! Maybe we can make it part of the release manager duties to update the timing files. I know it might be a little too often (every 2 months I think?) but otherwise I'm worried it will be forgotten.

side note for core peeps: while running the full CI when merging to master is definitely desirable and great, we might miss it. ESPECIALLY because merge commits _always_ fail CI (particularly, the docker builds). That's because it only builds the non-GPU, non-interface docker images if I'm following correctly, so it doesn't install TF, which means it installs the latest numpy, which means all SPSA (and other finite-diff?) gradient tests fail. we might wanna do something about it

**Comment by trbromley** [comment]
> But if we are okay with using more workers for lightened test runs, we can easily re-enable them. Let me know.

I'll defer to @albi3ro @mlxd here, but I feel a bit more confident if they are run.

> while running the full CI when merging to master is definitely desirable and great, we might miss it

How about adding it to the [test matrix](https://github.com/PennyLaneAI/plugin-test-matrix)?

**Comment by mlxd** [comment]
> > But if we are okay with using more workers for lightened test runs, we can easily re-enable them. Let me know.
> 
> I'll defer to @albi3ro @mlxd here, but I feel a bit more confident if they are run.
> 
> > while running the full CI when merging to master is definitely desirable and great, we might miss it
> 
> How about adding it to the [test matrix](https://github.com/PennyLaneAI/plugin-test-matrix)?

Yea, let's keep things enabled if possible. Unless we are stuck, I'd say it is worth the penalty.

**Pull request merged**
